src/vm/analysis/type_checker/natives/mod.rs

```diff
@@ -190,31 +190,53 @@ fn check_special_if(checker: &mut TypeChecker, args: &[SymbolicExpression], cont
 
 fn check_contract_call(checker: &mut TypeChecker, args: &[SymbolicExpression], context: &TypingContext) -> TypeResult {
     check_arguments_at_least(2, args)?;
-    let contract_identifier = match args[0].expr {
-        SymbolicExpressionType::LiteralValue(Value::Principal(PrincipalData::Contract(ref contract_identifier))) => contract_identifier,
-        _ => return Err(CheckError::new(CheckErrors::ContractCallExpectName))
-    };
 
-    let function_name = args[1].match_atom()
+    let func_name = args[1].match_atom()
         .ok_or(CheckError::new(CheckErrors::ContractCallExpectName))?;
     checker.type_map.set_type(&args[1], no_type())?;
 
-    let contract_call_function_type = {
-        if let Some(function_type) = checker.db.get_public_function_type(&contract_identifier, function_name)? {
-            Ok(function_type)
-        } else if let Some(function_type) = checker.db.get_read_only_function_type(&contract_identifier, function_name)? {
-            Ok(function_type)
-        } else {
-            Err(CheckError::new(CheckErrors::NoSuchPublicFunction(contract_identifier.to_string(),
-                                                                  function_name.to_string())))
-        }
-    }?;
-
-    let contract_call_args = checker.type_check_all(&args[2..], context)?;
-    
-    let return_type = contract_call_function_type.check_args(&contract_call_args)?;
+    let (expected_args, expected_returns) = match &args[0].expr {
+        SymbolicExpressionType::LiteralValue(Value::Principal(PrincipalData::Contract(ref contract_identifier))) => {
+            // Static dispatch
+            let contract_call_function = {
+                if let Some(FunctionType::Fixed(function)) = checker.db.get_public_function_type(&contract_identifier, func_name)? {
+                    Ok(function)
+                } else if let Some(FunctionType::Fixed(function)) = checker.db.get_read_only_function_type(&contract_identifier, func_name)? {
+                    Ok(function)
+                } else {
+                    Err(CheckError::new(CheckErrors::NoSuchPublicFunction(contract_identifier.to_string(),
+                                                                          func_name.to_string())))
+                }
+            }?;
+
```

Comment:
It looks like the call to `.check_args()` got dropped?  Also, is the `.type_check_all()` adequately replaced by the loop over each argument with `type_check_expects()`?

Reply:
After talking with @kantai, we want to support constructions like this:
```
    let contract_defining_trait = 
        "(define-trait trait-1 (
            (get-1 (uint) (response uint uint))))";
    let dispatching_contract =
        "(use-trait trait-1 .defun.trait-1)
        (define-public (wrapped-get-1 (contract <trait-1>)) 
            (contract-call? contract get-1 u0))";
    let impl_contract = 
        "(impl-trait .defun.trait-1)
        (define-public (get-1 (x uint)) (ok u1))";
    let caller_contract =
        "(define-public (foo-bar)
           (contract-call? .dispatch wrapped-get-1 .implem))"; 
```
meaning we also want to be able to pass literals (`.implem`) in a cascade of `contract-call`.
As such, we need to statically infer, using the context, that `.implem` is implementing a given trait (knowing that .implem could be implementing multiple traits)